### PR TITLE
Retry 407 crawlera errors with exp backoff up to 10 times

### PR DESCRIPTION
**Pull request by hcoura**
Some people have been asking for this for a while now.

My main concern with my current implementation is that if someone is actually having authentication issues (like adding the wrong API key to the settings) they will only be warned of that after 10 retries with exp backoff...

### CODE REVIEW

**Comment by hcoura** [comment]
Still missing docs ofc. But wanted to get discussion going

File: scrapy_crawlera/middleware.py [hunk 1]
```diff
@@ -191,6 +208,14 @@ def process_exception(self, request, exception, spider):
             self._clear_dns_cache()
             self._set_custom_delay(request, self.connection_refused_delay)
 
+    def _retry_auth(self, response, request):
+        logging.debug("Retrying crawlera request for authentication issue")
```
Comment (akshayphilar) [inline scrapy_crawlera/middleware.py]: @hcoura given how infrequently this occurs in production could we not change this to a **warning**? This should address your concerns around alerting users about the possibility of invalid credentials.

Comment (ejulio) [reply scrapy_crawlera/middleware.py]: I'd say to change it to, at least info, but warning looks better.

I think that, before going to production, people will run it locally at least once.
So, maybe they won't wait 10 minutes, but just check the log messages.

File: scrapy_crawlera/middleware.py [hunk 2]
```diff
@@ -191,6 +208,14 @@ def process_exception(self, request, exception, spider):
...
+        retries = response.meta.get('crawlera_auth_retry_times', 0) + 1
+        retryreq = request.copy()
+        retryreq.meta['crawlera_auth_retry_times'] = retries
+        retryreq.dont_filter = True
```
Comment (ejulio) [inline scrapy_crawlera/middleware.py]: Maybe you could increment some kind of stats for auth issues similar to the RetryMiddleware
https://github.com/scrapy/scrapy/blob/master/scrapy/downloadermiddlewares/retry.py#L82

File: tests/test_crawlera.py [hunk 3]
```diff
@@ -511,3 +511,75 @@ def test_noslaves_delays(self, random_uniform_patch):
...
+        default_delay = 0
+
+        self.settings['CRAWLERA_BACKOFF_STEP'] = backoff_step
+        self.settings['CRAWLERA_BACKOFF_MAX'] = max_delay
+
+        self.spider.crawlera_enabled = True
+        crawler = self._mock_crawler(self.spider, self.settings)
+        mw = self.mwcls.from_crawler(crawler)
+        mw.open_spider(self.spider)
+        mw.max_auth_retry_times = 4
```
Comment (ejulio) [inline tests/test_crawlera.py]: Maybe this could be a setting

**Comment by hcoura** [comment]
@ejulio answering all your comments here.

This is not a feature, this is a workaround a crawlera shortcoming. From that perspective I believe it should:

* Have no configuration
* Do not break spiders
* Fix the issue for everyone just by updating the package

So that's why I haven't added any settings to this. Or made it a separate middleware or anything. @akshayphilar was the one pushing for this, so I will be happy to hear his inputs as well.

About stats I really don't think this would happen outside crawlera restarts to make any stats relevant, but sure we could add if you guys feel like it's needed.

File: scrapy_crawlera/middleware.py (hunk 1 above)
Comment (hcoura) [reply scrapy_crawlera/middleware.py]: warning should be fine.

**Comment by akshayphilar** [comment]
Hey @ejulio, this is to address node restarts or any condition that could result in incorrect `407` responses from Crawlera. This has had impact on users and has required multiple projects to address this with their own separate middlewares. This PR should help alleviate those problems right from within the scrapy_crawlera library.

**Pull request merged**